### PR TITLE
Assignment 5

### LINKED ISSUE #11: A5: Match users

### PR DESCRIPTION
**Pull request by kflim**
- [x] Added RabbitMQ message queueing service
- [x] Added frontend for making request and checking status of request
- [x] Refactor local storage of requests in server to database storage of requests
- [x] Containerize the application

amqplib Channel API reference: https://amqp-node.github.io/amqplib/channel_api.html

### CODE REVIEW

File: src/pages/collab/index.tsx [hunk 1]
```diff
@@ -0,0 +1,247 @@
+import { useEffect, useRef, useState } from "react";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faXmark } from "@fortawesome/free-solid-svg-icons";
+
+import { api } from "~/utils/api";
+
+import LoadingIcon from "../../components/collab/LoadingIcon";
+
+const MatchRequestPage = () => {
+  const [difficulty, setDifficulty] = useState(-1);
```
Comment (Gabau) [inline src/pages/collab/index.tsx]: I think this is too many hooks to read and not really needed when we can do something like the following
<https://blog.logrocket.com/using-react-usestate-object/>

File: src/styles/globals.css [hunk 2]
```diff
@@ -2,3 +2,127 @@
 @tailwind components;
 @tailwind utilities;
 
+.lds-dual-ring {
```
Comment (Gabau) [inline src/styles/globals.css]: I see alot of conflicts with #16

File: package.json [hunk 3]
```diff
@@ -15,6 +17,11 @@
     "docker:down": "docker-compose down"
   },
   "dependencies": {
+    "@fortawesome/fontawesome-svg-core": "^6.4.2",
```
Comment (Gabau) [inline package.json:20]: Rmb to remove unneeded dependencies

**Review by Gabau — COMMENTED** [review]
Some comments

File: tsconfig.json [hunk 4]
```diff
@@ -9,7 +9,7 @@
     "forceConsistentCasingInFileNames": true,
     "noEmit": true,
     "esModuleInterop": true,
-    "module": "esnext",
+    "module": "CommonJS",
```
Comment (Gabau) [inline tsconfig.json]: Whats this?

File: src/pages/collab/index.tsx (hunk 1 above)
Comment (kflim) [reply src/pages/collab/index.tsx]: Refactored in new commit

File: src/styles/globals.css (hunk 2 above)
Comment (kflim) [reply src/styles/globals.css]: I imported the styles and it seemed fine

File: package.json (hunk 3 above)
Comment (kflim) [reply package.json:20]: Removed in new commit

File: tsconfig.json (hunk 4 above)
Comment (kflim) [reply tsconfig.json]: This one is for running the typescript server file alone. I reverted this change and overridden the compiler options for running the typescript file

File: src/styles/globals.css (hunk 2 above)
Comment (Gabau) [reply src/styles/globals.css]: As in #16 is using sass, not too sure if there is a clean way to resolve this, prob port this over to a sep sass file

File: package.json (hunk 3 above)
Comment (Gabau) [reply package.json:20]: we still using ws and tls?

Comment (Gabau) [reply package.json:20]: I think we should add them when we need them. At the very least not in this pr

File: tsconfig.json (hunk 4 above)
Comment (Gabau) [reply tsconfig.json]: Is there a need for running the server file alone tho

File: package.json (hunk 3 above)
Comment (kflim) [reply package.json:20]: Ok I can remove them for this pr

File: tsconfig.json (hunk 4 above)
Comment (kflim) [reply tsconfig.json]: What would you suggest? Another way would be to make a server class and move all the server logic there and create the server in one of the ts files

Comment (Gabau) [reply tsconfig.json]: Hmm, why can't we put the code in matchingServer in the matchingRouter?

File: src/components/collab/UserRequestHandler.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
+import amqp from "amqplib/callback_api";
```
Comment (Gabau) [inline src/components/collab/UserRequestHandler.ts]: Idt this belongs here? Prob in utils or something? Actually idt this code is used in the front end isit? I think this belongs in server

**Review by Gabau — COMMENTED** [review]
Some comments. Don't quite understand the amqp stuff yet tho

File: src/components/collab/UserRequestHandler.ts (hunk 5 above)
Comment (kflim) [reply src/components/collab/UserRequestHandler.ts]: Yeah I'll put it in the server file since no other file is using it

File: tsconfig.json (hunk 4 above)
Comment (kflim) [reply tsconfig.json]: If the server code runs in the router directly, I believe it won't allow any other code to run in the router while the amqp server code is running, feel free to test it tho. Plus, the server won't start until you call one of the router methods (when you click the button to queue)

Comment (Gabau) [reply tsconfig.json]: We could use a separate thread for this? https://threads.js.org/usage

Comment (Gabau) [reply tsconfig.json]: not sure if its worth tho. What is the use of matchingServer.ts? Haven't had time to fully understand it

Comment (kflim) [reply tsconfig.json]: **TLDR**: It acts as our server for the amqp messages

What it does is that it creates a channel with queues that can receive messages from client channels. This acts as our 'server' and we can send messages (matchRequests, cancelRequests etc) to it. Then, it performs matching every second and when a match is found, it sends messages back to the client channels using their own respective queues, indicating that a match has been found. For now, the message is just the id(random uuid for now) of the other person matched with them and whether the request was successful or not. There is a timeout set to 30 seconds, where the request sent will be removed and the message sent to the respective client channel will indicate unsuccessful match. This is basically the gist of it.

For reference, I have included the page on RPC for rabbitmq, which will probably help a lot in visualizing it: 
https://www.rabbitmq.com/tutorials/tutorial-six-javascript.html

Comment (Gabau) [reply tsconfig.json]: Am I right if I say that we are sending the trpc requests and then the trpc server will send an rpc to amqp? Seems odd

Comment (kflim) [reply tsconfig.json]: Yep, you can't run the amqp code on the browser (at least from what I can tell), so I chose to use the router to send the message to the amqp 'server'

https://stackoverflow.com/questions/45936495/uncaught-typeerror-webpack-require-connect-is-not-a-function-at-connec

File: src/server/matchingServer.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
+import amqp from "amqplib/callback_api";
+
+import { PrismaClient as PrismaClientPostgres } from "@prisma-db-psql/client";
+
+const MIN_DIFFICULTY = 0;
+const MAX_DIFFICULTY = 5;
+
+const prismaPostgres = new PrismaClientPostgres();
+
+const removeRequest = async (
```
Comment (Gabau) [inline src/server/matchingServer.ts]: Shouldn't this be handled in the cancelRequest within the trpc procedure? That way we don't need the prisma adaptor here. Maybe the prisma clear should also be handled on server startup

File: tsconfig.json (hunk 4 above)
Comment (Gabau) [reply tsconfig.json]: > If the server code runs in the router directly, I believe it won't allow any other code to run in the router while the amqp server code is running, feel free to test it tho. Plus, the server won't start until you call one of the router methods (when you click the button to queue)

Just tried to dump all the code in matchingServer into matchingRequest and ran next dev and it just work. Is there some specific workflow in which there is an error?

**Review by Gabau — COMMENTED** [review]
Seems to work fine.
Find the multiple rpc calls sketchy tho.

The docker file I used for testing
```yml
services:
  rabbitmq:
    image: rabbitmq:3-management-alpine
    container_name: 'rabbitmq'
    ports:
        - 5672:5672
        - 15672:15672
    volumes:
        - ~/.docker-conf/rabbitmq/data/:/var/lib/rabbitmq/
        - ~/.docker-conf/rabbitmq/log/:/var/log/rabbitmq
    networks:
        - rabbitmq_go_net
  postgres_db:
    image: postgres:12.16
    container_name: postgresdb-dev
    restart: always
    environment:
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_DB: "db"
    ports:
      - "5432:5432"
    volumes:
      - ./data/postgres:/var/lib/postgresql/data
  mongo_db:
    image: mongo:7.0.0
    container_name: mongodb-dev
    command: ["--replSet", "rs0", "--bind_ip_all"]
    restart: always
    # environment:
    #   MONGO_INITDB_ROOT_USERNAME: ${MONGO_USER}
    #   MONGO_INITDB_ROOT_PASSWORD: ${MONGO_PASSWORD}
    ports:
      - "27017:27017"
    volumes:
      - ./data/mongo:/data/db
      - ./.docker/mongo/rs-initiate.js:/docker-entrypoint-initdb.d/rs-initiate.js
      # - ./.docker/mongo/replica.key:/data/replica.key
  s3-bucket:
    build: ./s3/
    restart: always
    ports:
      - 9000:9000
    environment:
      S3_BUCKET_NAME: ${S3_BUCKET_NAME}
    volumes:
      - "./data/s3:/home/app/public/peer-prep"
networks:
  rabbitmq_go_net:
    driver: bridge
```

File: src/server/matchingServer.ts (hunk 6 above)
Comment (kflim) [reply src/server/matchingServer.ts]: Sure, I can move it to the router

File: tsconfig.json (hunk 4 above)
Comment (kflim) [reply tsconfig.json]: Hm it didn't work earlier for me for some reason, I'll refactor the server file later then

File: src/styles/globals.css (hunk 2 above)
Comment (kflim) [reply src/styles/globals.css]: I moved all the styling into the sass file in the new commit

**Comment by Gabau** [comment]
Whats left for this PR?

File: src/pages/collab/index.tsx [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+const MatchRequestPage = () => {
+  const [pageState, setPageState] = useState({
+    difficulty: -1,
+    category: "",
+    id: "",
+    response: "",
+    statusMessage: "Searching for partner...",
+    isWaiting: false,
+    requestFailed: false,
+    hasSetDifficulty: false,
```
Comment (Gabau) [inline src/pages/collab/index.tsx]: I think we can do away with these fields? The same check can be done by
```ts
difficulty != -1 && category != ""

```

File: src/pages/collab/index.tsx [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    hasSetCategory: false,
+    isDifficultyMissing: true,
```
Comment (Gabau) [inline src/pages/collab/index.tsx]: What's the difference between hasSetCategory and isCategoryMissing? Aren't the two just not of each other?

File: src/pages/collab/index.tsx [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+      <div className="absolute left-1/2 top-1/2 flex h-80 w-64 -translate-x-1/2 -translate-y-1/2 flex-col justify-evenly rounded-xl border p-5">
+        <div className="difficulty-menu">
+          <span className="choose">Choose Difficulty</span>
+
+          <div className="dropdown" onClick={onDifficultySelect}>
+            <div className="select text-center">
+              <span>Select Difficulty</span>
+            </div>
+            <ul className="dropdown-menu">
+              <li value={0} onClick={(event) => selectDifficulty(event)}>
```
Comment (Gabau) [inline src/pages/collab/index.tsx]: https://legacy.reactjs.org/docs/lists-and-keys.html
We can do this with a map?

File: package-lock.json
Comment (Gabau) [inline package-lock.json:1]: I think we are using yarn so this shouldn't be added?

File: src/server/api/trpc.ts [hunk 10]
```diff
@@ -38,11 +38,12 @@ interface CreateContextOptions {
  *
  * @see https://create.t3.gg/en/usage/trpc#-serverapitrpcts
  */
+
 const createInnerTRPCContext = (opts: CreateContextOptions) => {
   return {
     session: opts.session,
-    prismaPostgres,
-    prismaMongo,
+    prismaPostgres: prismaPostgres,
```
Comment (Gabau) [inline src/server/api/trpc.ts]: I don't think this change is needed?

**Review by Gabau — CHANGES_REQUESTED** [review]
LGTM, some minor issues.
Also realised we have to demonstrate two users.
image

I think we should make the matching service a protected route and restrict each user to one matching? Maybe in a separate PR cause this PR may get too large

File: src/pages/collab/index.tsx (hunk 7 above)
Comment (kflim) [reply src/pages/collab/index.tsx]: the hasSet fields are for the error messages, they only appear if the user hasnt selected/put any value for the field yet and the user clicks on the match button, which will set these fields to true and any change to the inputs before or after will set the hasSet fields to true if it isnt already true

File: src/pages/collab/index.tsx (hunk 8 above)
Comment (kflim) [reply src/pages/collab/index.tsx]: Yea I can remove that, its unnecessary

File: package-lock.json
Comment (kflim) [reply package-lock.json:1]: Will remove in next commit

File: src/server/api/trpc.ts (hunk 10 above)
Comment (kflim) [reply src/server/api/trpc.ts]: Yea this is an oops, I'll change it back

File: src/pages/collab/index.tsx (hunk 9 above)
Comment (kflim) [reply src/pages/collab/index.tsx]: Will refactor in next commit

File: src/pages/collab/index.tsx (hunk 7 above)
Comment (Gabau) [reply src/pages/collab/index.tsx]: If we need such a functionality we could make do with just one `hasSubmitted` field to do the check?
Could also considering using select element instead of just  a div as we can set the required flag so we don't have to handle this kind of logic. Its not that impt of a change tho
https://www.w3schools.com/tags/att_select_required.asp

Comment (kflim) [reply src/pages/collab/index.tsx]: True, I refactored to hasSubmitted, I think the div is better for styling purposes but we can change it later

**Review by Gabau — APPROVED** [review]
LGTM

**Review by Nafeij — APPROVED** [review]
LGTM

File: src/pages/collab/index.tsx [hunk 11]
```diff
@@ -179,6 +203,45 @@ const MatchRequestPage = () => {
     console.log("join session");
   };
 
+  const router = useRouter();
+  const {
+    data: userData,
+    isLoading: isLoadingUserData,
+    error: errorFetchingUser,
+  } = api.user.getCurrentUser.useQuery();
```
Comment (Gabau) [inline src/pages/collab/index.tsx]: For this why can't we use useSession? With that we also don't need to handle the loading
Will have to rmb to add the check after all the hooks tho cause it can lead rendering with more hooks if not careful
https://next-auth.js.org/getting-started/client#usesession

File: src/server/api/routers/matchRequest.ts [hunk 12]
```diff
@@ -211,67 +225,73 @@ const matchRequests = async (ch: amqp.Channel) => {
   }
 };
 
-amqp.connect("amqp://localhost", (err, conn) => {
-  if (err) throw err;
-
-  conn.createChannel((err, ch) => {
+void Promise.resolve(
```
Comment (Gabau) [inline src/server/api/routers/matchRequest.ts:228]: Why is this needed?

**Review by Gabau — COMMENTED** [review]
Some comments
